File: quick-sort/quickSort.js
function recursionSum(arr){
  if (arr.length == 0){
    return 0
  } else {
   return arr[0] + recursionSum(arr.slice(1))
  }

}

function itemsListCount(arr){
  if (arr.length == 0){
    return 0
  } else {
    return 1 + itemsListCount(arr.slice(1))
  }
}

function maxNumber(arr){
  if (arr.length === 0){
    return null
  }

  else if (arr.length === 1){
    return arr[0];
  }
  else {
    max = maxNumber(arr.slice(1));
    if(arr[0] > max ){
      return arr[0]
    }else{
      return max
    }
  }
}


function quickSort(arr) {
  if (arr.length < 2){
    return arr;
  }
  else {
    const pivot = arr[0];
    let less = arr.filter(n => n < pivot);
    let greater = arr.filter(n => n > pivot); 

    return quickSort(less).concat([pivot], quickSort(greater));
  }
}

const my_arr = [-1000,2,3,4,100,6,7, 30];

console.log(quickSort(my_arr));

File: quick-sort/quick_sort.rb
def sum(arr)
  if arr.empty?
    return 0 
  else
     arr[0] + sum(arr[1..-1])
  end
end

def items_list_count(arr)
  if arr.empty?
    return 0
  else
    1 + items_list_count(arr[1..-1])
  end
end

def max_number(arr)
  return nil if arr.empty?
  if arr.length === 1
    return arr[0]
  else 
     max = max_number(arr[1..-1])
     if arr[0] > max
      return arr[0]
     else
      return max
     end

  end

end

def quickSort(arr)

  if arr.length < 2
    return arr
  else  
    pivot = arr[0]
    r = arr[1..-1]
    less = r.select {|n| n < pivot}
    greater = r.select {|n| n > pivot}
  
    quickSort(less) + [pivot] + quickSort(greater);
  end


end

my_arr = [100, 2, 20, 4, 5]



puts(quickSort(my_arr))
 